include/boost/url/impl/url_base.ipp

```diff
@@ -1337,6 +1337,166 @@ end() const noexcept ->
+    if( d == 0 )
+        return first;
+    BOOST_ASSERT(d > 0);
+    int c = 0;
+    for( auto i = v.s_ + first.off_, e = v.s_ + last.off_; i != e; ++i )
+        c += (*i == '/');
+    BOOST_ASSERT(c > 0);
+    BOOST_ASSERT(v.pt_.nseg >= c);
+    v.pt_.nseg -= c;
+    std::memmove(v.s_ + first.off_, v.s_ + last.off_, v.pt_.offset[detail::id_end] - last.off_ + 1);
+    v.pt_.resize(detail::id_path, v.pt_.length(detail::id_path, detail::id_query) - d);
+    BOOST_ASSERT(v.size() + d == v.a_.size());
+    v.a_.resize(v.size());
+    BOOST_ASSERT(v.s_[v.pt_.offset[detail::id_end]] == '\0');
+    first.parse();
+    return first;
+}
+
+auto
+url_base::
+segments_type::
+insert_encoded_segment_( segments_type::iterator pos, string_view s ) ->
+    iterator
+{
+    BOOST_ASSERT(detail::pchar_pct_set().check(s));
+    BOOST_ASSERT(v_ != nullptr);
+    url_base & v = *v_;
+    BOOST_ASSERT(v.size() == v.a_.size());
+    BOOST_ASSERT(pos.v_ == &v);
+    BOOST_ASSERT(pos.off_ >= v.pt_.offset[detail::id_path]);
+    BOOST_ASSERT(pos.off_ <= v.pt_.offset[detail::id_query]);
+    auto const n0 = v.pt_.offset[detail::id_end];
+    auto const n = s.size() + 1;
+    v.s_ = v.a_.resize(v.size() + n);
+    v.pt_.resize(detail::id_path, v.pt_.length(detail::id_path, detail::id_query) + n);
+    std::memmove(v.s_ + v.pt_.offset[detail::id_end] + pos.off_ - n0, v.s_ + pos.off_, n0 - pos.off_ + 1);
+    BOOST_ASSERT(v.s_[v.pt_.offset[detail::id_end]] == '\0');
+    v.s_[pos.off_] = '/';
+    std::memcpy(v.s_ + pos.off_ + 1, s.data(), s.size());
+    ++v.pt_.nseg;
+    pos.off_ += n;
+    pos.parse();
+    return pos;
+}
+
+auto
+url_base::
+segments_type::
+insert_segment_( segments_type::iterator pos, string_view s, std::size_t const ns ) ->
+    iterator
+{
+    BOOST_ASSERT(v_ != nullptr);
+    url_base & v = *v_;
+    BOOST_ASSERT(v.size() == v.a_.size());
+    BOOST_ASSERT(pos.v_ == &v);
+    BOOST_ASSERT(pos.off_ >= v.pt_.offset[detail::id_path]);
+    BOOST_ASSERT(pos.off_ <= v.pt_.offset[detail::id_query]);
+    auto const pct = detail::pchar_pct_set();
+    BOOST_ASSERT(pct.encoded_size(s) == ns);
+    auto const n0 = v.pt_.offset[detail::id_end];
+    auto const n = ns + 1;
+    v.s_ = v.a_.resize(v.size() + n);
+    v.pt_.resize(detail::id_path, v.pt_.length(detail::id_path, detail::id_query) + n);
+    std::memmove(v.s_ + v.pt_.offset[detail::id_end] + pos.off_ - n0, v.s_ + pos.off_, n0 - pos.off_ + 1);
+    BOOST_ASSERT(v.s_[v.pt_.offset[detail::id_end]] == '\0');
+    v.s_[pos.off_] = '/';
+    pct.encode(v.s_ + pos.off_ + 1, s);
+    ++v.pt_.nseg;
+    pos.off_ += n;
+    pos.parse();
+    return pos;
+}
+
+auto
+url_base::
+segments_type::
+insert_encoded_segment( segments_type::iterator pos, string_view s ) ->
+    iterator
+{
+    return insert_encoded_segment_(pos, detail::pchar_pct_set().validate(s));
+}
+
+auto
+url_base::
+segments_type::
+insert_segment( segments_type::iterator pos, string_view s ) ->
```

Comment:
do we need the suffix `_segment` on every function? It seems redundant

Reply:
There is encoded/not encoded. Do you prefer `insert`/`insert_encoded`?

Reply:
I think having to type `_segment` every time is burdensome. Note that _every_ function you added has it, which suggests it is redundant.

Reply:
Agreed. Changing it.